sdks/python/apache_beam/ml/inference/api.py

```diff
@@ -0,0 +1,84 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from dataclasses import dataclass
+import apache_beam as beam
+from typing import Tuple, TypeVar, Union
+# TODO: implement RunInferenceImpl
+# from apache_beam.ml.inference.base import RunInferenceImpl
+
+
+@dataclass
+class BaseModelSpec:
+  model_url: str
+
+
+@dataclass
+class PyTorchModel(BaseModelSpec):
+  device: str
+
+  def __post_init__(self):
+    self.device = self.device.upper()
+
+
+@dataclass
+class SklearnModel(BaseModelSpec):
+  serialization_type: str
+
+  def __post_init__(self):
+    self.serialization_type = self.serialization_type.upper()
+
+
+_K = TypeVar('_K')
+_INPUT_TYPE = TypeVar('_INPUT_TYPE')
+_OUTPUT_TYPE = TypeVar('_OUTPUT_TYPE')
+
+
+@dataclass
+class PredictionResult:
+  key: _K
+  example: _INPUT_TYPE
+  inference: _OUTPUT_TYPE
+
+
+@beam.ptransform_fn
+@beam.typehints.with_input_types(Union[_INPUT_TYPE, Tuple[_K, _INPUT_TYPE]])
+@beam.typehints.with_output_types(PredictionResult)
```

Comment:
How should we deal with typing? TFX [explicitly](https://github.com/tensorflow/tfx-bsl/blob/v1.6.0/tfx_bsl/public/beam/run_inference.py#L34) defines output as `Union[_OUTPUT_TYPE, Tuple[_K, _OUTPUT_TYPE]]`

Reply:
yeah, TFX is doing it right I think.  They either have output type or a key with output type.

I'm not the typing expert, IMO its ok to have a typing specific PR that just concentrates on that too, if this gets messy.

We need typing to support TFX's proto use case as well as ours new frameworks use case.

Reply:
I've modified the output type and pushed the change. I feel like we should try to get typing as accurate as possible early on, especially if we want to ensure that we can accommodate TFX's implementation.

Could we have, instead of a bare `PredictionResult` type, a parent class from which `PredictionResult` and `prediction_log_pb2.PredictionLog` inherit? Or use a `Union`? (but this may get messy).

This definitely requires input from someone with more experience with typing.